contracts/oracles.cairo

```diff
@@ -0,0 +1,96 @@
+%lang starknet
+
+from starkware.cairo.common.pow import pow
+from starkware.cairo.common.math_cmp import is_le
+from starkware.cairo.common.math import unsigned_div_rem
+from starkware.cairo.common.bool import TRUE, FALSE
+
+from contracts.Math64x61 import (
+    Math64x61_fromFelt,
+    Math64x61_div,
+    Math64x61_INT_PART,
+    Math64x61_add,
+    Math64x61_sqrt,
+)
+
+from contracts._cfg import EMPIRIC_ORACLE_ADDRESS, EMPIRIC_AGGREGATION_MODE
+
+# List of available tickers:
+#  https://docs.empiric.network/using-empiric/supported-assets
+
+# Contract interface copied from docs
+@contract_interface
+namespace IEmpiricOracle:
+    func get_value(key : felt, aggregation_mode : felt) -> (
+        value : felt, decimals : felt, last_updated_timestamp : felt, num_sources_aggregated : felt
+    ):
+    end
+end
+
+# Function for iterative division
+func iter_div_64x61{range_check_ptr}(x : felt, y : felt) -> (res : felt):
+    # both x and y are Math64x61
+    # this is needed because of weird error in overflow
+    # this may introduce imprecisions
+    alloc_locals
+
+    let (pow_10_to_30) = pow(10, 30)
+    let (is_convertable) = is_le(y, pow_10_to_30)
+
+    if is_convertable == TRUE:
+        let (res_a) = Math64x61_div(x, y)
+        return (res_a)
+    end
+
+    let (div_a) = Math64x61_sqrt(y)
+    let (div_b) = Math64x61_div(y, div_a)
+
+    let (partial_res) = iter_div_64x61(x, div_a)
+    let (res) = iter_div_64x61(partial_res, div_b)
+
+    return (res)
+end
+
+# Function to convert base 10**18 number from oracle to base 2**61
+# which is used throughout the AMM
+func convert_price{range_check_ptr}(price : felt, m : felt) -> (price : felt):
+    alloc_locals
+
+    let (is_convertable) = is_le(price, Math64x61_INT_PART)
+    if is_convertable == TRUE:
+        let (converted_price) = Math64x61_fromFelt(price)
+        let (pow10xM) = pow(10, m)
+        let (pow10xM_to_64x61) = Math64x61_fromFelt(pow10xM)
+        let (price_64x61) = iter_div_64x61(converted_price, pow10xM_to_64x61)
+
+        return (price_64x61)
+    end
+
+    let (m_1, r) = unsigned_div_rem(m, 2)
+    let m_2 = m - m_1
+
+    let (pow_10_m1) = pow(10, m_1)
+    let (c, remainder) = unsigned_div_rem(price, pow_10_m1)
+
+    let (a) = convert_price(c, m_2)
+    let (b) = convert_price(remainder, m)
+
+    let (res) = Math64x61_add(a, b)
+
+    # FIXME: THIS HAS TO VALIDATED THAT THE ROUNDING CAUSED BY THE IMPRECISE CALCULATIONS IS NOT TOO BIG
+
+    return (res)
+end
+
+@view
+func empiric_median_price{syscall_ptr : felt*, range_check_ptr}(key : felt) -> (price : felt):
```

Comment:
is it really median price? I'm not sure what you are getting from empiric... if so, could write somewhere which different metrics of price you can get from empiric?